### PR TITLE
feat: configurable disk caching

### LINKED ISSUE #2249: Enable configuration of disk caching type
Currently we statically set the disk caching type to `"ReadWrite"` for all OS disks, and `"ReadOnly"` for all data disks.

The azure-sdk-for-go defines three caching types:

```
const (
	// CachingTypesNone ...
	CachingTypesNone CachingTypes = "None"
	// CachingTypesReadOnly ...
	CachingTypesReadOnly CachingTypes = "ReadOnly"
	// CachingTypesReadWrite ...
	CachingTypesReadWrite CachingTypes = "ReadWrite"
)
```

Let's enable user-configurable caching types (validate against the three sane values above) per-masterProfile, and per-agentPoolProfile, for both OS disks and data disks.

### PR DESCRIPTION
**Pull request by jackfrancis**
**Reason for Change**:

This PR introduces configurable disk caching:

For OS disk configuration on master VMs:
- either "ReadOnly", "ReadWrite", or "None"

For OS disk configuration on agent pool VMs:
- either "ReadOnly", "ReadWrite", or "None"
For data disk configuration on agent pool VMs:
- either "ReadOnly", "ReadWrite", or "None"
  - supports a single configuration option across all data disks

An example api model w/ this new configuration:

```
{
...
        "masterProfile": {
            ...
            "osDiskCachingType": "ReadOnly",
            ...
        },
        "agentPoolProfiles": [
            {
            ...
            "osDiskCachingType": "ReadWrite",
            "dataDiskCachingType": "None",
            ...
            }
        ]
...
}
```

**Issue Fixed**:

Fixes #2249 

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

**Comment by jackfrancis** [comment]
@PacoGuy How does the api model spec in the PR description look in terms of what you'd want in a configurable disk caching configuration?

And @mboersma FYI on intent, and @marosset to make sure we ship this w/ Windows 1st class support as well.

**Comment by PacoGuy** [comment]
Looks like that PR model spec would work for us, Thanks!

**Comment by marosset** [comment]
@jackfrancis This should only add changes to the ARM correct?
If so I don't think anything specific needs to be done from windows (other than adding a test config or something to verify/validate)
LMK if you do need help tho.

File: examples/e2e-tests/kubernetes/release/default/definition.json [hunk 1]
```diff
@@ -121,7 +123,8 @@
         "vmSize": "Standard_D2s_v3",
         "osType": "Windows",
         "availabilityProfile": "VirtualMachineScaleSets",
-        "vnetSubnetId": "/subscriptions/SUB_ID/resourceGroups/RG_NAME/providers/Microsoft.Network/virtualNetworks/VNET_NAME/subnets/SUBNET_NAME"
+        "vnetSubnetId": "/subscriptions/SUB_ID/resourceGroups/RG_NAME/providers/Microsoft.Network/virtualNetworks/VNET_NAME/subnets/SUBNET_NAME",
+        "osDiskCachingType": "ReadOnly"
```
Comment (jackfrancis) [inline examples/e2e-tests/kubernetes/release/default/definition.json:127]: @marosset FYI, adding this new configuration to a Windows pool in our E2E cluster configuration to confirm that this is not a Linux/Windows thing.

Btw, should we add a data disk to this pool to (1) test that functionality generally, and to also validate the data disk caching type override?

File: pkg/api/vlabs/validate_test.go [hunk 2]
```diff
@@ -2881,6 +2882,39 @@ func TestMasterProfileValidate(t *testing.T) {
...
+			masterProfile: MasterProfile{
+				DNSPrefix:         "baz",
+				Count:             3,
+				OSDiskCachingType: string(compute.CachingTypesReadOnly),
+			},
+		},
+		{
+			name: "Master Profile with invalid OSDiskCachingType",
+			masterProfile: MasterProfile{
+				DNSPrefix:         "whizbang",
```
Comment (mboersma) [inline pkg/api/vlabs/validate_test.go:2912]: I thought "qux" followed "baz." 😜

Comment (jackfrancis) [reply pkg/api/vlabs/validate_test.go:2912]: There is no spoon.

File: examples/e2e-tests/kubernetes/release/default/definition.json (hunk 1 above)
Comment (marosset) [reply examples/e2e-tests/kubernetes/release/default/definition.json:127]: Sure.

**Pull request merged**